pi5/hw_pwm.go

```diff
@@ -166,6 +195,99 @@ func (pwm *pwmDevice) callGPIOReadall() {
 	}
 }
 
+// Helper Function for SetPinMode. This method updates the given mode of a pin by finding its specific location in memory & writing to the 'mode' byte in the 8 byte block of pin data.
+func (pwm *pwmDevice) writeToPinModeByte(GPIONumber int, newMode byte) error {
+
+	/*
+		Of the 8 bytes representing all of a given pin's data:
+			pinBytes[0:3] -> bytes are allocated for 'status' modes (still unsure of what status does).
+			pinBytes[4:7] -> bytes are allocated for alternative modes.
+
+		However, you only need 1 byte to represent all the modes because
+			1 byte = 8 bits
+			8 modes can be represented by 3 bits
+
+		In practice, only 1 of the mode bytes is changed and the rest are just 0xff's.
+		The index of that byte in the entire chunk of pin data is pinBytes[4].
+	*/
+
+	altModeIndex := 4
+
+	pinAddress, err := getGPIOPinAddress(GPIONumber)
+	if err != nil {
+		return fmt.Errorf("error getting gpio pin address: %w", err)
+	}
+
+	// find pin data within virtual page; retrieve the 4th byte from the pin data
+	vPage := *(pwm.gpioChipVPage)
+	pinBytes := vPage[pinAddress : pinAddress+pinDataSize]
+	altModeByte := pinBytes[altModeIndex]
+
+	// We keep the left half (4 bits) of the byte preserved, only modifying the right half
+	// Preserve Left Side of Byte using this Mask
+	leftSideMask := byte(0xf0)
+
+	// Set Right Side of Byte ONLY using this Mask. Ensures left side cannot be overwritten
+	rightSideMask := byte(0x0f)
+
+	// Set new mode via write to correct while protecting previous other settings
+	newAltModeByte := (altModeByte & leftSideMask) | (newMode & rightSideMask)
+	pinBytes[altModeIndex] = newAltModeByte
+
+	return nil
+}
+
+/*
+For all pins belonging to the same bank, pin data is stored contiguously and in 8 byte chunks.
+For a given pin, this method determines:
+ 1. which bank the pin belongs to
+ 2. the starting address of its 8 byte data chunk
+*/
+func getGPIOPinAddress(GPIONumber int) (int64, error) {
+
```

Comment:
Mild preference to never have a blank line after a `{`. We're about to change indentation levels; we don't also need to start a new paragraph to indicate that we're starting something new.